Refactor data fetching API docs (#30615)
vercel/next.js: docs/api-reference/data-fetching/getStaticPaths.md

---

> description: Fetch data at build time with `getStaticProps`.

Reviewer: ```suggestion
description: Fetch data and generate static pages with `getStaticProps`.
```

> description: Fetch data and generate static pages with `getStaticProps`.

Reviewer: ```suggestion
description: Fetch data and generate static pages with `getStaticPaths`.
```

---

# `getStaticPaths`

<details>
  <summary><b>Version History</b></summary>

| Version  | Changes                                                                                                           |
| -------- | ----------------------------------------------------------------------------------------------------------------- |

> | `v9.5.0` | Stable [Incremental Static Regeneration](https://nextjs.org/blog/next-9-5#stable-incremental-static-regeneration) |

Reviewer: Should we update this to the ISR docs?

Author: Good catch!

> | `v9.5.0` | Stable [Incremental Static Regeneration](/docs/basic-features/data-fetching/incremental-static-regeneration.md) |

Reviewer: Follow up PR - let's add more updates here, we've definitely added more stuff between here and 9.5

Author: https://github.com/vercel/next.js/issues/31953

> | `v9.3.0` | `getStaticPaths` introduced.                                                                                    |

Reviewer 2: Could not format this table properly in the review UI, but a new line should be added
```
| `v10.0.0` | `fallback: 'blocking'` return option added.
```

https://nextjs.org/blog/next-10#blocking-fallback-for-getstaticpaths

Author: I created a separate issue for updating these

</details>

> When exporting an `async` function called `getStaticPaths` from a page that uses dynamic routes, Next.js will statically pre-render all the paths specified by `getStaticPaths`.

Reviewer: Linkify dynamic routes?

> When exporting an `async` function called `getStaticPaths` from a page that uses [dynamic routes](/docs/routing/dynamic-routes.md), Next.js will statically pre-render all the paths specified by `getStaticPaths`.

Reviewer: ```suggestion
When exporting an `async` function called `getStaticPaths` from a page that uses [Dynamic Routes](/docs/routing/dynamic-routes.md), Next.js will statically pre-render all the paths specified by `getStaticPaths`.
```

```jsx
export async function getStaticPaths() {
  return {
    paths: [
      { params: { ... } } // See the "paths" section below
    ],

>     fallback: true or false // See the "fallback" section below

Reviewer 2: ```suggestion
    fallback: true, false or "blocking" // See the "fallback" section below
```

  };
}
```

## `getStaticPaths` return values

The `getStaticPaths` function should return an object with the following **required** properties:

### `paths`

> The `paths` key determines which paths will be pre-rendered. For example, suppose that you have a page that uses dynamic routes named `pages/posts/[id].js`. If you export `getStaticPaths` from this page and return the following for `paths`:

Reviewer: ```suggestion
The `paths` key determines which paths will be pre-rendered. For example, suppose that you have a page that uses [Dynamic Routes](/docs/routing/dynamic-routes.md) named `pages/posts/[id].js`. If you export `getStaticPaths` from this page and return the following for `paths`:
```

```js
return {
  paths: [
    { params: { id: '1' } },
    { params: { id: '2' } }
  ],
  fallback: ...
}
```

> Then Next.js will statically generate `posts/1` and `posts/2` at build time using the page component in `pages/posts/[id].js`.

Reviewer: ```suggestion
Then, Next.js will statically generate `/posts/1` and `/posts/2` during `next build` using the page component in `pages/posts/[id].js`.
```

> Note that the value for each `params` must match the parameters used in the page name:

Reviewer: ```suggestion
The value for each `params` object must match the parameters used in the page name:
```

- If the page name is `pages/posts/[postId]/[commentId]`, then `params` should contain `postId` and `commentId`.

> - If the page name uses catch-all routes, for example `pages/[...slug]`, then `params` should contain `slug` which is an array. For example, if this array is `['foo', 'bar']`, then Next.js will statically generate the page at `/foo/bar`.

Reviewer: ```suggestion
- If the page name uses [catch-all routes](/docs/routing/dynamic-routes.md#catch-all-routes) like `pages/[...slug]`, then `params` should contain `slug` (which is an array). If this array is `['hello', 'world']`, then Next.js will statically generate the page at `/hello/world`.
```

Reviewer: Let's avoid foo/bar 🙏

Author: OMG i didnt add that, and missed it completely, thanks for spotting this! 100% do not want foo/bar in the docs!

> - If the page uses an optional catch-all route, supply `null`, `[]`, `undefined` or `false` to render the root-most route. For example, if you supply `slug: false` for `pages/[[...slug]]`, Next.js will statically generate the page `/`.

Reviewer: ```suggestion
- If the page uses an [optional catch-all route](/docs/routing/dynamic-routes.md#optional-catch-all-routes), use `null`, `[]`, `undefined` or `false` to render the root-most route. For example, if you supply `slug: false` for `pages/[[...slug]]`, Next.js will statically generate the page `/`.
```

What is "root most route"?

### `fallback: false`

> If `fallback` is `false`, then any paths not returned by `getStaticPaths` will result in a **404 page**. You can do this if you have a small number of paths to pre-render - so they are all statically generated during build time.

Reviewer: ```suggestion
If `fallback` is `false`, then any paths not returned by `getStaticPaths` will result in a **404 page**. This is useful if you have a small number of paths to pre-render and want to statically generate them all during `next build`.
```

> It is also useful when the new pages are not added often. If you add more items to the data source and need to render the new pages, you will need to run the build again.

Reviewer: How could we make this more clear? Could be a follow up PR

> When `next build` is run, Next.js will check if `getStaticPaths` returned `fallback: false`, it will then build **only** the paths returned by `getStaticPaths`. This option is useful if you have a small number of paths to create, or new page data is not added often. If you find that you need to add more paths, and you have `fallback: false`, you will need to run `next build` again so that the new paths can be generated.

Reviewer 2: "This is useful if you have a small number of paths to pre-render..."

"This option is useful if you have a small number of paths to create, or new page data is not added often..."

create or pre-render?

Personally, one of these sentences/sections feels like repetition.

  // Render post...
}

// This function gets called at build time
export async function getStaticPaths() {
  // Call an external API endpoint to get posts
  const res = await fetch('https://.../posts')
  const posts = await res.json()

  // Get the paths we want to pre-render based on posts
  const paths = posts.map((post) => ({
    params: { id: post.id },
  }))

  // We'll pre-render only these paths at build time.
  // { fallback: false } means other routes should 404.
  return { paths, fallback: false }
}

// This also gets called at build time
export async function getStaticProps({ params }) {
  // params contains the post `id`.
  // If the route is like /posts/1, then params.id is 1
  const res = await fetch(`https://.../posts/${params.id}`)
  const post = await res.json()

  // Pass post data to the page via props
  return { props: { post } }
}

export default Post
```

### `fallback: true`

<details>
  <summary><b>Examples</b></summary>
  <ul>
    <li><a href="https://static-tweet.vercel.app">Static generation of a large number of pages</a></li>
  </ul>
</details>

If `fallback` is `true`, then the behavior of `getStaticProps` changes in the following ways:

> - The paths returned from `getStaticPaths` will be rendered to `HTML` at build time by `getStaticProps`.

Reviewer 2: "rendered to `HTML`" is a new term in addition to "path to create/pre-render"?

Author: I disagree, the paths returned _are_ rendered to HTML. The other section which talks about paths to create, is not talking about whats rendered after the creation of the path.

- The paths that have not been generated at build time will **not** result in a 404 page. Instead, Next.js will serve a [“fallback”](#fallback-pages) version of the page on the first request to such a path.
- In the background, Next.js will statically generate the requested path `HTML` and `JSON`. This includes running `getStaticProps`.
- When complete, the browser receives the `JSON` for the generated path. This will be used to automatically render the page with the required props. From the user’s perspective, the page will be swapped from the fallback page to the full page.
- At the same time, Next.js adds this path to the list of pre-rendered pages. Subsequent requests to the same path will serve the generated page, just like other pages pre-rendered at build time.

> **Note:** `fallback: true` is not supported when using [`next export`](/docs/advanced-features/static-html-export.md).

#### When is `fallback: true` useful?

> `fallback: true` is useful if your app has a very large number of static pages that depend on data (such as a very large e-commerce site). If you want to pre-render all product pages, the builds would take forever.

Reviewer 2: ```suggestion
`fallback: true` is useful if your app has a very large number of static pages that depend on data (such as a very large e-commerce site). If you want to pre-render all product pages, the builds would take a very long time.
```

Let's tone this down :sweat_smile:

Instead, you may statically generate a small subset of pages and use `fallback: true` for the rest. When someone requests a page that is not generated yet, the user will see the page with a loading indicator or skeleton component.

Shortly after, `getStaticProps` finishes and the page will be rendered with the requested data. From now on, everyone who requests the same page will get the statically pre-rendered page.

This ensures that users always have a fast experience while preserving fast builds and the benefits of Static Generation.

`fallback: true` will not _update_ generated pages, for that take a look at [Incremental Static Regeneration](/docs/basic-features/data-fetching/incremental-static-regeneration).

> ### `fallback: 'blocking'`

Reviewer 2: I believe this whole section is less important since:

https://nextjs.org/blog/next-12#bot-aware-isr-fallback
https://github.com/vercel/next.js/discussions/28180

Should somehow incorporate that extra information.

---
description: Fetch data and generate static pages with `getStaticProps`.
---

# `getStaticPaths`

<details>
  <summary><b>Version History</b></summary>

| Version  | Changes                                                                                                      |
| -------- | ------------------------------------------------------------------------------------------------------------ |
| `v9.5.0` | Stable [Incremental Static Regeneration](/docs/basic-features/data-fetching/incremental-static-regeneration) |
| `v9.3.0` | `getStaticPaths` introduced.                                                                                 |

</details>

> When exporting an `async` function called `getStaticPaths` from a page that uses [dynamic routes](/docs/routing/dynamic-routes), Next.js will statically pre-render all the paths specified by `getStaticPaths`.

Reviewer: ```suggestion
When exporting an `async` function called `getStaticPaths` from a page that uses [dynamic routes](/docs/routing/dynamic-routes.md), Next.js will statically pre-render all the paths specified by `getStaticPaths`.
```
